File: js/src/maplestory/api/msea/dto/character/characterVMatrix.ts
import * as base from '../../../common/dto/character/characterVMatrix';
import {
  CharacterVMatrixBody,
  CharacterVMatrixCoreEquipmentDtoBody,
} from '../../response/character/characterVMatrixBody';

/**
 * Character V Matrix information
 */
export class CharacterVMatrixDto extends base.CharacterVMatrixDto {
  /**
   * Reference date for query (SGT, daily data with hours and minutes set to 0)
   */
  public override date: Date | null;

  /**
   * Character job
   */
  public override characterClass: string | null;

  /**
   * V Matrix Node information
   */
  public override characterVCoreEquipment: CharacterVMatrixCodeEquipmentDto[];

  /**
   * Remaining Matrix enhancement points for the character
   */
  public override characterVMatrixRemainSlotUpgradePoint: number | null;

  constructor(obj: CharacterVMatrixBody) {
    super();

    const {
      date,
      character_class,
      character_v_core_equipment,
      character_v_matrix_remain_slot_upgrade_point,
    } = obj;

    this.date = date ? new Date(date) : null;
    this.characterClass = character_class;
    this.characterVCoreEquipment = character_v_core_equipment.map(
      (equipment) => new CharacterVMatrixCodeEquipmentDto(equipment),
    );
    this.characterVMatrixRemainSlotUpgradePoint =
      character_v_matrix_remain_slot_upgrade_point;
  }
}

/**
 * V Matrix Node information
 */
export class CharacterVMatrixCodeEquipmentDto extends base.CharacterVMatrixCodeEquipmentDto {
  /**
   * Slot index
   */
  public override slotId: string;

  /**
   * Slot level
   */
  public override slotLevel: number;

  /**
   * Node name
   */
  public override vCoreName: string | null;

  /**
   * Node type
   */
  public override vCoreType: string | null;

  /**
   * Node level
   */
  public override vCoreLevel: number;

  /**
   * Name of the skill that corresponds to the node
   */
  public override vCoreSkill1: string;

  /**
   * Name of the second skill that corresponds to the node (for Boost Nodes)
   */
  public override vCoreSkill2: string | null;

  /**
   * Name of the third skill that corresponds to the node (for Boost Nodes)
   */
  public override vCoreSkill3: string | null;

  constructor(obj: CharacterVMatrixCoreEquipmentDtoBody) {
    super();

    const {
      slot_id,
      slot_level,
      v_core_name,
      v_core_type,
      v_core_level,
      v_core_skill_1,
      v_core_skill_2,
      v_core_skill_3,
    } = obj;

    this.slotId = slot_id;
    this.slotLevel = slot_level;
    this.vCoreName = v_core_name;
    this.vCoreType = v_core_type;
    this.vCoreLevel = v_core_level;
    this.vCoreSkill1 = v_core_skill_1;
    this.vCoreSkill2 = v_core_skill_2;
    this.vCoreSkill3 = v_core_skill_3;
  }
}

File: csharp/MapleStory.OpenAPI/Src/MSEA/Dto/Character/CharacterVMatrix.cs
using Newtonsoft.Json;
using Base = MapleStory.OpenAPI.Common.DTO;

namespace MapleStory.OpenAPI.MSEA.DTO
{
    /// <summary>
    /// Character V Matrix information
    /// </summary>
    public class CharacterVMatrixDTO : Base.CharacterVMatrixDTO<CharacterVMatrixCoreEquipmentDTO>
    {
        /// <summary>
        /// Reference date for query (SGT, daily data with hours and minutes set to 0)
        /// </summary>
        [JsonProperty("date")]
        public override DateTimeOffset? Date
        {
            get => _date?.ToOffset(TimeSpan.FromHours(8));
            set => _date = value;
        }

        private DateTimeOffset? _date;

        /// <summary>
        /// Character job
        /// </summary>
        [JsonProperty("character_class")]
        public override string CharacterClass { get; set; }

        /// <summary>
        /// V Matrix Node information
        /// </summary>
        [JsonProperty("character_v_core_equipment")]
        [JsonConverter(typeof(NullAsEmptyConverter))]
        public override List<CharacterVMatrixCoreEquipmentDTO> CharacterVCoreEquipment { get; set; } = new();

        /// <summary>
        /// Remaining Matrix enhancement points for the character
        /// </summary>
        [JsonProperty("character_v_matrix_remain_slot_upgrade_point")]
        public override int CharacterVMatrixRemainSlotUpgradePoint { get; set; }
    }

    /// <summary>
    /// V Matrix Node information
    /// </summary>
    public class CharacterVMatrixCoreEquipmentDTO : Base.CharacterVMatrixCoreEquipmentDTO
    {
        /// <summary>
        /// Slot index
        /// </summary>
        [JsonProperty("slot_id")]
        public override string SlotId { get; set; }

        /// <summary>
        /// Slot level
        /// </summary>
        [JsonProperty("slot_level")]
        public override long SlotLevel { get; set; }

        /// <summary>
        /// Node name
        /// </summary>
        [JsonProperty("v_core_name")]
        public override string? VCoreName { get; set; }

        /// <summary>
        /// Node type
        /// </summary>
        [JsonProperty("v_core_type")]
        public override string? VCoreType { get; set; }

        /// <summary>
        /// Node level
        /// </summary>
        [JsonProperty("v_core_level")]
        public override long VCoreLevel { get; set; }

        /// <summary>
        /// Name of the skill that corresponds to the node
        /// </summary>
        [JsonProperty("v_core_skill_1")]
        public override string VCoreSkill1 { get; set; }

        /// <summary>
        /// Name of the second skill that corresponds to the node (for Boost Nodes)
        /// </summary>
        [JsonProperty("v_core_skill_2")]
        public override string? VCoreSkill2 { get; set; }

        /// <summary>
        /// Name of the third skill that corresponds to the node (for Boost Nodes)
        /// </summary>
        [JsonProperty("v_core_skill_3")]
        public override string? VCoreSkill3 { get; set; }
    }
}
